File: simple_robot_py/src/obstacle_detector.py
#!/usr/bin/env python3

# obstacle_detector.py
# Author: Mohamad Moudallal
# GitHub: https://github.com/moudallal
# Email: mohamad@moudallal.xyz

import rospy
from sensor_msgs.msg import LaserScan
from simple_robot.msg import Obstacle

# Subscriber callback method
def callback(data):
    # Access the forward laser scan distance (0-2pi -> ranges[0]-ranges[359])
    range = []
    range.extend(list(data.ranges[355:360]))
    range.extend(list(data.ranges[0:5]))

    # Send the minimum distance in a 10-degree range for ensuring safety
    obstacle.distance = min(range)
    # Send a dummy name
    obstacle.name = 'Obstacle'
    publisher.publish(obstacle)

# Initialize ROS node
rospy.init_node('obstacle_detector', anonymous=True)
# Define publisher on topic /obstacle_detection
publisher = rospy.Publisher('/obstacle_detection', Obstacle, queue_size=10)
# Define subscriber on topic /scan
subscriber = rospy.Subscriber('/scan', LaserScan, callback)

# Set rate to 10Hz
r = rospy.Rate(10)
# Define the message to be published
obstacle = Obstacle()

# Keep the node in spin
rospy.spin()

File: simple_robot_cpp/src/obstacle_detector.cpp
// obstacle_detector.cpp
// Author: Mohamad Moudallal
// GitHub: https://github.com/moudallal
// Email: mohamad@moudallal.xyz

#include "ros/ros.h"
#include <vector>
#include <algorithm>

#include <sensor_msgs/LaserScan.h>
#include "simple_robot_msgs/Obstacle.h"

// Initialize the publisher and subscriber
ros::Publisher publisher;
ros::Subscriber subscriber;

// Subscriber callback method
void callback(const sensor_msgs::LaserScan data) {
    // Access the forward laser scan distance (0-2pi -> ranges[0]-ranges[359])
    std::vector<float> range;
    range.insert(range.begin(), data.ranges.begin() + 354, data.ranges.end());
    range.insert(range.end(), data.ranges.begin(), data.ranges.begin() + 5);
    
    simple_robot_msgs::Obstacle obstacle;
    // Send the minimum distance in a 10-degree range for ensuring safety
    obstacle.distance = *std::min_element(range.begin(), range.end());
    // Send a dummy name
    obstacle.name = "Obstacle";
    publisher.publish(obstacle);
}

int main(int argc, char **argv) {
    // Initialize a ROS node
    ros::init(argc, argv, "obstacle_detector");

    // Create a ROS NodeHandle object
    ros::NodeHandle n;

    // Define publisher on topic /obstacle_detection
    publisher = n.advertise<simple_robot_msgs::Obstacle>("/obstacle_detection", 10);
    // Define subscriber on topic /scan
    subscriber = n.subscribe<sensor_msgs::LaserScan>("/scan", 10, callback);

    // Keep the node in spin
    ros::spin();

    return 0;
}